src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java

```diff
@@ -72,32 +78,68 @@ private boolean shouldHaveConnectedByNow(T c) {
     }
 
     @Override
-    protected void kill(Node n) throws IOException {
-        MansionComputer computer = (MansionComputer) n.toComputer();
-        // attempt to heuristically detect a race condition
-        // where an executor accepted a task after we checked for
-        // idleness,
-        // but before we marked it as unavailable for tasks
-        // See JENKINS-23676
-        LOGGER.log(Level.FINE, "Taking node offline since it seems to be idle" + n.getNodeName());
+    protected void kill(final Node n) throws IOException {       
+        final MansionComputer computer = (MansionComputer) n.toComputer();
+        
+        final String nodeName = n.getNodeName();
+        LOGGER.log(Level.FINE, "Taking node {0} offline since it seems to be idle", n.getNodeName());
+        // we need to have a private block as other threads could try to turn on accepting tasks
+        computer.setDisconnectInProgress(true);
+        // set accepting tasks so that others can see this flag if expecting it
         computer.setAcceptingTasks(false);
-        try {
-            Thread.sleep(2000);
-        } catch (InterruptedException e) {
-            LOGGER.log(Level.FINE, "Interrupted while sleeping before removing node " + n.getNodeName());
-            return;
-        }
-        if (!computer.isIdle() && computer.isOnline()) {
-            LOGGER.log(FINE, computer.getName() + " is no longer idle, aborting termination.");
-            // we lost the race -- mark it as back online
-            computer.setAcceptingTasks(true);
-            return;
-        }
-        for (Executor e : computer.getExecutors()) {
-            e.interrupt();
+        synchronized (this) {
+            if (disconnectInProgress) {
+                // if we already have a background thread running, then we can return immediately
+                return;
+            }
+            disconnectInProgress = true;
         }
-        LOGGER.log(Level.FINE, "Finally removing node " + n.getNodeName());
-        super.kill(n);
+        Computer.threadPoolForRemoting.submit(new Runnable() {
+            public void run() {
+                // TODO once Jenkins 1.607+ we can remove the sleep as the withLock will give atomic guarantee
+
+                // attempt to heuristically detect a race condition
+                // where an executor accepted a task after we checked for
+                // idleness,
+                // but before we marked it as unavailable for tasks
+                // See JENKINS-23676
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    LOGGER.log(Level.FINE, "Interrupted while sleeping before removing node {0}", nodeName);
+                    return;
+                }
+                
```

Comment:
Now that we have queueLock.withLock() it seems to me like this sleep hack is no longer required. Can't you remove this entire try block?